metrics-core/src/lib.rs

```diff
@@ -32,24 +32,172 @@
 //!
 //! Histograms are a convenient way to measure behavior not only at the median, but at the edges of
 //! normal operating behavior.
+#![deny(missing_docs)]
 use futures::future::Future;
 use std::borrow::Cow;
+use std::fmt;
+use std::slice::Iter;
 use std::time::Duration;
 
-/// An optimized metric key.
+/// An allocation-optimized string.
 ///
-/// As some metrics might be sent at high frequency, it makes no sense to constantly allocate and
-/// reallocate owned [`String`]s when a static [`str`] would suffice.  As we don't want to limit
-/// callers, though, we opt to use a copy-on-write pointer -- [`Cow`] -- to allow callers
-/// flexiblity in how and what they pass.
-pub type Key = Cow<'static, str>;
+/// We specify `ScopedString` to attempt to get the best of both worlds: flexibility to provide a
+/// static or dynamic (owned) string, while retaining the performance benefits of being able to
+/// take ownership of owned strings and borrows of completely static strings.
+pub type ScopedString = Cow<'static, str>;
 
-/// A value which can be converted into a nanosecond representation.
+/// A key/value pair used to further describe a metric.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct Label(ScopedString, ScopedString);
+
+impl Label {
+    /// Creates a `Label` from a key and value.
+    pub fn from_parts<K, V>(key: K, value: V) -> Self
+    where
+        K: Into<ScopedString>,
+        V: Into<ScopedString>,
+    {
+        Label(key.into(), value.into())
+    }
+}
+
+/// A metric key.
+///
+/// A key always includes a name, but can optional include multiple labels used to further describe
+/// the metric.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct Key {
+    #[doc(hidden)]
+    pub name: ScopedString,
```

Comment:
why hidden public?

Reply:
Good call; I created the type originally thinking I'd use the fields, and then I decided on accessors, but never cleaned up the fields.